isis/src/kaguya/objs/KaguyaTcCamera/KaguyaTcCameraDistortionMap.cpp

```diff
@@ -0,0 +1,207 @@
+ *   http://www.usgs.gov/privacy.html.
+ */
+
+#include <QDebug>
+#include <QtGlobal>
+#include <QtMath>
+
+#include "KaguyaTcCameraDistortionMap.h"
+
+namespace Isis {
+  /** 
+   * Kaguya TC Camera distortion map constructor
+   *
+   * Create a camera distortion map for Kaguya's TC1 and TC2
+   * This class maps between distorted and undistorted 
+   * focal plane x/y's. The default mapping is the identity, that is, 
+   * the focal plane x/y and undistorted focal plane x/y will be 
+   * identical. 
+   *
+   * @param parent        the parent camera that will use this distortion map
+   * @param zDirection    the direction of the focal plane Z-axis
+   *                      (either 1 or -1)
+   *
+   */
+  KaguyaTcCameraDistortionMap::KaguyaTcCameraDistortionMap(Camera *parent, int naifIkCode)
+      : CameraDistortionMap(parent) {
+    QString odtxkey = "INS" + toString(naifIkCode) + "_DISTORTION_COEF_X";
+    QString odtykey = "INS" + toString(naifIkCode) + "_DISTORTION_COEF_Y";
+    
+    for(int i = 0; i < 4; ++i) {
+      p_odkx.push_back(p_camera->getDouble(odtxkey, i));
+      p_odky.push_back(p_camera->getDouble(odtykey, i));
+    }
+  }    
+
+
+
+  /**
+   * Destructor
+   */
+  KaguyaTcCameraDistortionMap::~KaguyaTcCameraDistortionMap() {
+  }
+
+
+  /** 
+   * Compute undistorted focal plane x/y
+   *
+   * Compute undistorted focal plane x/y given a distorted focal plane x/y.
+   * This virtual method can be used to apply various techniques for removing
+   * optical distortion in the focal plane of a camera.  The default
+   * implementation uses a polynomial distortion if the SetDistortion method
+   * is invoked.  After calling this method, you can obtain the undistorted
+   * x/y via the UndistortedFocalPlaneX and UndistortedFocalPlaneY methods
+   *
+   * @param dx distorted focal plane x in millimeters
+   * @param dy distorted focal plane y in millimeters
+   *
+   * @return if the conversion was successful
+   * @see SetDistortion
+   * @todo Generalize polynomial equation
+   */
+  bool KaguyaTcCameraDistortionMap::SetFocalPlane(double dx, double dy) {
+    p_focalPlaneX = dx;
+    p_focalPlaneY = dy;
+
+    double x = dx;
+    double y = dy;
+
+    double r2 = x*x + y*y;
+    double r = qSqrt(r2); 
+    double r3 = r2 * r; 
+
+    // Apply distortion correction
+    // see: SEL_TC_V01.TI
+    // r2 = x^2 + y^2
+    //   Line-of-sight vector of pixel no. n can be expressed as below.
+
+  //  Distortion coefficients information:
+  //   INS<INSTID>_DISTORTION_COEF_X  = ( a0, a1, a2, a3)
+  //   INS<INSTID>_DISTORTION_COEF_Y  = ( b0, b1, b2, b3),
+  //
+  // Distance r from the center:
+  //   r = - (n - INS<INSTID>_CENTER) * INS<INSTID>_PIXEL_SIZE.
+  //
+  // Line-of-sight vector v is calculated as
+  //   v[X] = INS<INSTID>BORESIGHT[X]
+  //          +a0 +a1*r +a2*r^2 +a3*r^3 ,
+  //   v[Y] = INS<INSTID>BORESIGHT[Y]
+  //          +r +a0 +a1*r +a2*r^2 +a3*r^3 ,
+  //   v[Z] = INS<INSTID>BORESIGHT[Z] .
```

Comment:
Is any of this needed?

Reply:
I think this is useful to have in here because it's the actual equation information that we implemented in this function.